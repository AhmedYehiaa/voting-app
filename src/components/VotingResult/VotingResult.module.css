.votingResult {
  border: 1px solid #ccc;
  width: 75%;
  margin: 50px auto;
}

.votingResult ul {
  list-style: none;
  padding: 0;
}

.votingResult li {
  padding: 0 20px;
}

.votingResult .title {
  margin: 0;
  padding: 10px;
  /* background: #002837; */
  /* color: #fff; */
}

.votingResult .progressBar {
  position: relative;
  height: 40px;
  line-height: 40px;
  margin-bottom: 50px;
  border-radius: 0;
  border: 1px solid #ccc;
  background-color: #fff;
  font-weight: 500;
  border-radius: 0.6rem;
  overflow: hidden;
}

.votingResult .filler {
  background: rgb(0, 41, 55);
  color: #fff;
  height: 100%;
  border-radius: 0.6rem 0 0 0.6rem;
  transition: width 0.2s ease-in;
}

.votingResult .filler__percentage {
  margin-left: 20px;
}
