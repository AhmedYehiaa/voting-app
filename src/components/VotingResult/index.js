import React from "react";
import PropTypes from "prop-types";

import styles from "./VotingResult.module.css";

const VotingResult = ({ question }) => {
  const getTotalVotes = choices => {
    return choices.reduce((sum, curr) => (sum += curr.votes), 0);
  };

  const totalVotes = getTotalVotes(question.choices);

  return (
    <div className={styles.votingResult}>
      <h1 className={styles.title}>{question.title}</h1>
      <ul>
        {question.choices.map(({ choice, url, votes }) => (
          <li key={url}>
            <p>
              {choice} ({votes} {"  "}
              {votes > 1 ? <span>Votes</span> : <span>Vote</span>})
            </p>
            <div className={styles.progressBar}>
              <div
                className={styles.filler}
                style={{
                  width: `${(votes / totalVotes) * 100}%`
                }}
              >
                <span className={styles.filler__percentage}>
                  {((votes / totalVotes) * 100).toFixed(2)}%
                </span>
              </div>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
};

VotingResult.propTypes = {
  title: PropTypes.string,
  question: PropTypes.shape({
    title: PropTypes.string.isRequired,
    choices: PropTypes.arrayOf(PropTypes.stirng)
  }).isRequired
};

export default VotingResult;
