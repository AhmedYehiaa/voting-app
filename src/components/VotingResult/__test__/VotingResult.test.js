import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

import VotingResult from "../index";

afterEach(cleanup);

describe("Voting Result Component", () => {
  const question = {
    title: "Favourite hot beverage?",
    published_at: "2015-05-27T21:22:26.648000+00:00",
    url: "/questions/9",
    choices: [
      {
        choice: "Apple Cider",
        votes: 30,
        url: "/questions/9/choices/67"
      },
      {
        choice: "Tea",
        votes: 30,
        url: "/questions/9/choices/65"
      },
      {
        choice: "Hot Chocolate",
        votes: 21,
        url: "/questions/9/choices/68"
      },
      {
        choice: "Coffee",
        votes: 17,
        url: "/questions/9/choices/66"
      }
    ]
  };

  it("render without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Router>
        <VotingResult question={question} />
      </Router>,
      div
    );
  });

  it("render Voting Result page with question", () => {
    const { getByTestId } = render(
      <Router>
        <VotingResult question={question} />
      </Router>
    );
    expect(getByTestId("title").textContent).toBe(question.title);
  });

  it("render choices list with count equals to 4", () => {
    const { getByTestId } = render(
      <Router>
        <VotingResult question={question} />
      </Router>
    );
    const choices = getByTestId("choices");
    expect(choices.children.length).toBe(4);
  });

  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Router>
          <VotingResult question={question} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
