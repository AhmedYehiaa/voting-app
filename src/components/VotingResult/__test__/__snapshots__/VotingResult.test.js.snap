// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Voting Result Component matches snapshot 1`] = `
<div
  className="votingResult"
  data-testid="votingResult"
>
  <h2
    className="title"
    data-testid="title"
  >
    Favourite hot beverage?
  </h2>
  <ul
    data-testid="choices"
  >
    <li>
      <p>
        Apple Cider
         (
        30
         
          
        <span>
          Votes
        </span>
        )
      </p>
      <div
        className="progressBar"
      >
        <div
          className="filler"
          style={
            Object {
              "width": "30.612244897959183%",
            }
          }
        >
          <span
            className="filler__percentage"
          >
            30.61
            %
          </span>
        </div>
      </div>
    </li>
    <li>
      <p>
        Tea
         (
        30
         
          
        <span>
          Votes
        </span>
        )
      </p>
      <div
        className="progressBar"
      >
        <div
          className="filler"
          style={
            Object {
              "width": "30.612244897959183%",
            }
          }
        >
          <span
            className="filler__percentage"
          >
            30.61
            %
          </span>
        </div>
      </div>
    </li>
    <li>
      <p>
        Hot Chocolate
         (
        21
         
          
        <span>
          Votes
        </span>
        )
      </p>
      <div
        className="progressBar"
      >
        <div
          className="filler"
          style={
            Object {
              "width": "21.428571428571427%",
            }
          }
        >
          <span
            className="filler__percentage"
          >
            21.43
            %
          </span>
        </div>
      </div>
    </li>
    <li>
      <p>
        Coffee
         (
        17
         
          
        <span>
          Votes
        </span>
        )
      </p>
      <div
        className="progressBar"
      >
        <div
          className="filler"
          style={
            Object {
              "width": "17.346938775510203%",
            }
          }
        >
          <span
            className="filler__percentage"
          >
            17.35
            %
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
